File: app/src/main/java/com/bignerdranch/stockwatcher/StockWatcherApplication.java
package com.bignerdranch.stockwatcher;

import android.app.Application;
import android.content.Context;

import lombok.Getter;

public class StockWatcherApplication extends Application {

    @Getter
    private AppComponent appComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        initAppComponent();
    }

    private void initAppComponent() {
        appComponent = DaggerAppComponent.builder()
                .appModule(new AppModule())
                .build();
    }

    public static StockWatcherApplication get(Context context) {
        return (StockWatcherApplication) context.getApplicationContext();
    }

    public static AppComponent getAppComponent(Context context) {
        StockWatcherApplication stockWatcherApplication = get(context.getApplicationContext());
        return stockWatcherApplication.getAppComponent();
    }

}

File: app/src/main/java/com/bignerdranch/stockwatcher/StockWatcherApplication.kt
package com.bignerdranch.stockwatcher

import android.app.Application
import android.content.Context

class StockWatcherApplication : Application() {

    private lateinit var appComponent: AppComponent

    override fun onCreate() {
        super.onCreate()
        initAppComponent()
    }

    private fun initAppComponent() {
        appComponent = DaggerAppComponent.builder()
                .appModule(AppModule())
                .build()
    }

    companion object {

        operator fun get(context: Context): StockWatcherApplication {
            return context.applicationContext as StockWatcherApplication
        }

        fun getAppComponent(context: Context): AppComponent {
            val stockWatcherApplication = get(context.applicationContext)
            return stockWatcherApplication.appComponent
        }
    }

}
